include/ibamr/MassIntegrator.h

```diff
@@ -0,0 +1,326 @@
+ * \f$ \nabla \cdot (\rho \mathbf{u} Q)\f$ for the energy equation where \f$ Q = C_\textrm{p} T\f$
+ * or \f$ Q = h\f$ based on the form of the energy equation.
+ *
+ * \note MassIntegrator is an intermediate or a lightweight integrator that is used within one time step to
+ * get a consistent mass flux \f$ m_\rho = \rho \mathbf{u} \f$ for advecting either momentum \f$ \mathbf{u} \f$
+ * through \f$ \nabla \cdot (\rho \mathbf{u} \otimes \mathbf{u})  = \nabla \cdot (m_\rho \otimes \mathbf{u} ) \f$
+ * or a scalar \f$ Q \f$ through  \f$ \nabla \cdot (\rho \mathbf{u} Q)  = \nabla \cdot (m_\rho  Q) \f$. It does
+ * not maintain any hierarchy data, child or parent integrator, as done by advection-diffusion hierarchy integrator.
+ *
+ * The core concept behind `MassIntegrator` is to obtain a consistent mass flux $\mathbf{m_\rho} = \rho \mathbf{u}$
+ * for stabilizing high density ratio flows. It also ensures a consistency condition that:
+ *
+ * 1. When there are no viscous, pressure forces or any other body forces in the domain, the momentum equation becomes
+ *    discretely equal to the mass equation.
+ * 2. Under isothermal/isenthalpic conditions, with no additional heat source, the energy/enthalpy equation reverts to
+ *    mass balance equation.
+ *
+ * To achieve these conditions, the time integrator scheme of MassIntegrator is tightly coupled to how
+ * INSVCStaggeredHierarchyIntegrator or PhaseChangeHierarchyIntegrator integrate their respective variables.
+ * Note that when flow is incompressible, one does not need to solve mass balance equation.
+ * Here, we are solving a redundant mass balance equation through MassIntegrator class in order to achieve consistency
+ * and stability.
+ *
+ * Reference
+ * Nangia et. al, <A HREF="https://www.sciencedirect.com/science/article/pii/S0021999119302256">
+ * A robust incompressible Navier-Stokes solver for high density ratio multiphase flows </A>
+ *
+ * Thirumalaisamy and Bhalla, <A HREF="https://arxiv.org/abs/2301.06256">
+ * A low Mach enthalpy method to model non-isothermal gas-liquid-solid flows with melting and solidification </A>
+ *
+ */
+class MassIntegrator : public virtual SAMRAI::tbox::DescribedClass
+{
+public:
+    /*!
+     * \brief Class constructor.
+     */
+    MassIntegrator(std::string object_name, SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db);
+
+    /*!
+     * \brief Destructor.
+     */
+    virtual ~MassIntegrator() = default;
+
+    /*!
+     * \brief Integrate density and compute the convective operator of the momentum and/or energy.
+     */
+    virtual void integrate(double dt) = 0;
+
+    /*!
+     * \name General operator functionality.
+     */
+    //\{
+
+    /*!
+     * \brief Compute hierarchy dependent data required for time integrating variables.
+     */
+    virtual void
+    initializeTimeIntegrator(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM> > base_hierarchy) = 0;
```

Comment:
If we go with "single time step integrator", suggest naming these:
```suggestion
    initializeSTSIntegrator(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM> > base_hierarchy) = 0;
```

Reply:
(I won't mark the others, but the same comment applies below.)